tools/signal.py

```diff
@@ -0,0 +1,447 @@
+        continious_s_down = binary_dilation(continious_s_down)
+    for i in range(noise_reduction):
+        continious_s_up = binary_erosion(continious_s_up)
+        continious_s_down = binary_erosion(continious_s_down)
+
+    return continious_s_up, continious_s_down
+
+
+def get_closest_signal_pair(
+    s1: np.ndarray,
+    s2: np.ndarray,
+    s1_value: int = -1,
+    s2_value: int = 1,
+) -> tuple[np.ndarray, np.ndarray]:
+    """The function returns the positions in s1 and s2 for where pairs of
+    patterns s1-value -> s2-value are found (s1-value is assumed to preceed
+    s2-value)."""
+
+    s1_positions = np.where(s1 == s1_value)[0]
+    s2_positions = np.where(s2 == s2_value)[0]
+
+    # Match all
+    signals: list[tuple[int, int]] = []
+    for p in s1_positions:
+        tmp_diff = s2_positions - p
+        tmp_diff = tmp_diff[tmp_diff > 0]
+        if tmp_diff.size > 0:
+            p2 = tmp_diff.min() + p
+            if len(signals) > 0 and p2 == signals[-1][1]:
+                if p2 - p < signals[-1][1] - signals[-1][0]:
+                    del signals[-1]
+                    signals.append((p, p2))
+            else:
+                signals.append((p, p2))
+
+    signals_arr = np.array(signals)
+    if signals_arr.size == 0:
+        raise SignalError("No signal pairs found")
+
+    return signals_arr[:, 0], signals_arr[:, 1]
+
+
+def get_signal_spikes(
+    down_slopes: np.ndarray,
+    up_slopes: np.ndarray,
+) -> np.ndarray:
+    """Returns where valleys are in a signal based on down and up slopes"""
+    # combined_signal = (
+    #     down_slopes.astype(int) * -1 + up_slopes.astype(int)
+    # )
+
+    # Edge-detect so that signal start is >0 and signal end <0
+    kernel = np.array([-1, 1])
+    d_down = np.round(
+        fftconvolve(down_slopes, kernel, mode='same'),
+    ).astype(int)
+    d_up = np.round(
+        fftconvolve(up_slopes, kernel, mode='same'),
+    ).astype(int)
+
+    s1, s2 = get_closest_signal_pair(d_up, d_down, s1_value=-1, s2_value=1)
+    return (s1 + s2) / 2.0
+
+
+"""
+def _get_closest(X: np.ndarray, Y: np.ndarray) -> np.ndarray:
+
+    new_list = []
+    for i in ideal_signal:
+        delta_i = np.abs(X - i)
+        delta_reciprocal = np.abs(Y - X[delta_i.argmin()])
+        if delta_i.min() == delta_reciprocal.min():
+            new_list += ([X[delta_i.argmin()],
+                Y[delta_reciprocal.argmin()]])
+
+    return np.array(new_list)
+"""
+
+
+def _get_alt_closest(X: np.ndarray, Y: np.ndarray) -> np.ndarray:
+    dist = np.abs(np.subtract.outer(X, Y))
+    idx1 = np.argmin(dist, axis=0)
+    idx2 = np.argmin(dist, axis=1)
+    Z = np.c_[
+        X[idx1[idx2] == np.arange(len(X))],
+        Y[idx2[idx1] == np.arange(len(Y))]
+    ].ravel()
+    return Z
+
+
+def _get_orphans(X: np.ndarray, shortX: np.ndarray) -> np.ndarray:
+    return X[np.abs(np.subtract.outer(X, shortX)).min(axis=1).astype(bool)]
+
+
+def get_offset_quality(
+    s,
```

Comment:
This and `w1` aren't very nice. Maybe @local-minimum has ideas for better names? (Here and below.)

Reply:
nope and the point of `/tools` is to not really maintain the code until we notice we or someone needs it so I would argue to put minimal time fixing things here... perhaps typos but not more.

Reply:
Is the plan in the long run, then, to remove this directory?

Reply:
move it to a new repo